src/gpu/kmeans/kmeans_h2o4gpu.cu

```diff
@@ -172,6 +178,456 @@ void random_centroids(int verbose, int seed, const char ord,
+    if (curr_prob >= best_prob) {
+      best_prob = curr_prob;
+      best_prob_idx = i;
+    }
+
+    cur_weight += (data_val * weights[i]);
+    i += 1;
+  }
+  return i > -1 ? i : best_prob_idx;
+}
+
+/**
+ * Copies cols records, starting at position idx*cols from data to centroids. Removes them afterwards from data.
+ * Removes record from weights at position idx.
+ * @tparam T
+ * @param idx
+ * @param cols
+ * @param data
+ * @param weights
+ * @param centroids
+ */
+template<typename T>
+void add_centroid(int idx, int cols,
+                  thrust::host_device<T> &data,
+                  thrust::host_device<T> &weights,
+                  thrust::host_device<T> &centroids) {
+  for (int i = 0; i < cols; i++) {
+    centroids.push_back(data[idx * cols + i]);
+  }
+  for (int i = 0; i < cols; i++) {
+    data.remove(idx * cols + i);
+  }
+  weights.remove(idx);
+}
+
+/**
+ * K-Means++ algorithm
+ * @tparam T
+ * @param seed
+ * @param data
+ * @param weights
+ * @param k
+ * @param cols
+ * @param centroids
+ */
+template<typename T>
+void kmeans_plus_plus(
+    int seed,
+    thrust::host_device<T> data,
+    thrust::host_device<T> weights,
+    int k,
+    int cols,
+    thrust::host_device<T> &centroids) {
+
+  int centroid_idx = pick_point_idx_weighted(
+      pairwise_cost,
+      weights
+  );
+
+  add_centroid(centroid_idx, cols, data, weights, centroids);
+
+  std::vector<T> best_pairwise_distances(data.size());
+  std::vector<T> std_data(data.data());
+  std::vector<T> std_centroids(centroids.data());
+
+  compute_distances(std_data,
+                    std_centroids,
+                    best_pairwise_distances,
+                    data.size(), cols, 1);
+
+  for (int iter = 0; iter < k; iter++) {
+    centroid_idx = pick_point_idx_weighted(
+        best_pairwise_distances,
+        weights
+    );
+
+    add_centroid(centroid_idx, cols, data, weights, centroids);
+
+    best_pairwise_distances.remove(centroid_idx);
+
+    // TODO necessary?
+    std_data = std::vector<T>(data.data());
+    std_centroids = std::vector<T>(centroids.data() + cols * (iter + 1));
+
+    std::vector<T> curr_pairwise_distances(data.size());
+
+    compute_distances(std_data,
+                      std_centroids,
+                      curr_pairwise_distances,
+                      std_data.size(), cols, 1);
+
+    for (int i = 0; i < curr_pairwise_distances.size(); i++) {
+      best_pairwise_distances[i] = std::min(curr_pairwise_distances[i], best_pairwise_distances[i]);
+    }
+  }
+}
+
+/**
+ * Calculates closest centroid for each record and counts how many points are assigned to each centroid.
+ * @tparam T
+ * @param verbose
+ * @param num_gpu
+ * @param rows_per_gpu
+ * @param cols
+ * @param data
+ * @param data_dots
+ * @param centroids
+ * @param weights
+ * @param pairwise_distances
+ * @param labels
+ */
+template<typename T>
+void count_pts_per_centroid(
+    int verbose,
+    int num_gpu, int rows_per_gpu, int cols,
```

Comment:
Prefer passing a reference to a vector of vectors to using pointer pointers here.